src/Raven.Server/Utils/Enumerators/TransactionForgetAboutAbstractEnumerator.cs

```diff
@@ -16,16 +16,27 @@ protected TransactionForgetAboutAbstractEnumerator([NotNull] IEnumerator<T> inne
         DocsContext = docsContext;
     }
 
-    protected abstract void ForgetAbout(T item);
+    protected abstract void ForgetAbout(ForgetAboutItem item);
+    protected abstract ForgetAboutItem CloneCurrent(T item);
+
+    protected struct ForgetAboutItem
+    {
+        public T Item;
+        public T CompressedItem;
+        public bool IsCloned;
+    }
 
     public bool MoveNext()
     {
-        ForgetAbout(Current);
+        ForgetAbout(_current);
 
         if (_innerEnumerator.MoveNext() == false)
             return false;
 
-        Current = _innerEnumerator.Current;
+        // the clone is needed because we are going to forget about the _current document
+        // Current should be disposed by the caller
+        _current = CloneCurrent(_innerEnumerator.Current);
```

Comment:
It will have significant memory usage implications, won't it? Especially in operations like streaming which is going to deal with large data sets

Reply:
why? 
1. for uncompressed we don't do anything
2. for compressed we do `Dispose` & `ForgetAbout` for original (compressed) bjro, and the item created by clone should be disposed by caller of the enumerator